src/main/java/com/ververica/flink/table/gateway/deployment/ClusterDescriptorAdapter.java

```diff
@@ -0,0 +1,161 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.ververica.flink.table.gateway.deployment;
+
+import com.ververica.flink.table.gateway.SqlExecutionException;
+import com.ververica.flink.table.gateway.SqlGatewayException;
+import com.ververica.flink.table.gateway.context.ExecutionContext;
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.api.common.JobStatus;
+import org.apache.flink.client.deployment.ClusterDescriptor;
+import org.apache.flink.client.program.ClusterClient;
+import org.apache.flink.configuration.Configuration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * Adapter to handle kinds of job actions (eg. get job status or cancel job) based on execution.target
+ */
+public abstract class ClusterDescriptorAdapter<ClusterID> {
+    private static final Logger LOG = LoggerFactory.getLogger(ClusterDescriptorAdapter.class);
+    private static final int DEFAULT_TIMEOUT_SECONDS = 30;
+
+    protected final ExecutionContext<ClusterID> executionContext;
+    // Only used for logging
+    private final String sessionId;
+    // jobId is not null only after job is submitted
+    protected final JobID jobId;
+    protected final Configuration configuration;
+    protected final ClusterID clusterID;
+
+    public ClusterDescriptorAdapter(ExecutionContext<ClusterID> executionContext,
+                                    Configuration configuration,
+                                    String sessionId,
+                                    JobID jobId
+                                    ) {
+        this.executionContext = executionContext;
+        this.sessionId = sessionId;
+        this.jobId = jobId;
+        this.configuration = configuration;
+        this.clusterID = executionContext.getClusterClientFactory().getClusterId(configuration);
+    }
+
+    /**
+     * Returns the status of the flink job.
+     */
+    public JobStatus getJobStatus(){
+        if (jobId == null) {
+            LOG.error("Session: {}. No job has been submitted. This is a bug.", sessionId);
+            throw new IllegalStateException("No job has been submitted. This is a bug.");
+        }
+        return bridgeClientRequest(this.executionContext, jobId, sessionId, clusterClient -> {
+				try {
+					return clusterClient.getJobStatus(jobId).get(DEFAULT_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+				} catch (InterruptedException | ExecutionException | TimeoutException e) {
+					LOG.error(String.format("Session: %s. Failed to fetch job status for job %s", sessionId, jobId), e);
+					throw new SqlGatewayException("Failed to fetch job status for job " + jobId, e);
+				}
```

Comment:
It is possible that when the job is almost finished, we can get the `clusterClient` successfully. But when we use `clusterClient` to query the status, the job finished and a `TimeoutException` will be thrown. Maybe we should shorten `DEFAULT_TIMEOUT_SECONDS` and add a `DEFAULT_RETRIES` (for example 3 times)?